matern kernel

gf712 opened the pull request:
This kernel will be useful for Bayesian optimisation.

gf712 committed:
matern kernel

gf712 committed:
clang fix

karlnapf reviewed (commented):
gradients will be necessary for tuning hyperparameters in BO :)

karlnapf commented:
whats needed to merge this?

gf712 commented:
> whats needed to merge this?

Gradients and unit tests :D

gf712 commented:
and general clean up. I'll see if I have some time later this week.

gf712 commented on src/shogun/kernel/MaternKernel.cpp:
```diff
@@ -0,0 +1,85 @@
+	float64_t dist = distance->distance(idx_a, idx_b);
+
+	// first we can check if we should use one of the approximations which are cheaper to calculate
+	if (CMath::fequals(nu, 0.5, std::numeric_limits<float64_t>::epsilon()))
```
@vigsterkr do we have this somewhere else in the library now?

karlnapf replied:
i dont think atm

gf712 commented on src/shogun/kernel/MaternKernel.cpp:
```diff
@@ -0,0 +1,85 @@
+#ifdef __clang__
+		// clang hasn't implemented mathematical special functions in the C++17 standard at this point
+		// note that yn casts nu to an integer
+		float64_t bessel = yn(nu, ratio);
```
need to check if new clang has this implemented

gf712 replied:
@karlnapf can we copy the implementation of bessel function of the second kind from https://www.boost.org/doc/libs/1_56_0/boost/math/special_functions/bessel.hpp? The problem is that clang doesn't have the implementation, even though it is part of the C++17 standard.

karlnapf replied:
yep, just copy paste it somewhere appropriate (check license)

gf712 commented on src/shogun/kernel/MaternKernel.cpp:
```diff
@@ -0,0 +1,85 @@
+
+#include <cmath>
+
+CMaternKernel::CMaternKernel(): CDistanceKernel(), nu(0.0)
```
move the default value to class member definition

gf712 commented on src/shogun/kernel/MaternKernel.h:
```diff
@@ -0,0 +1,72 @@
+			CMaternKernel(int32_t size, float64_t width, float64_t order, CDistance* dist);
+
+			/** destructor */
+			~CMaternKernel();
```
can be default, and needs override

gf712 commented on src/shogun/kernel/MaternKernel.h:
```diff
@@ -0,0 +1,72 @@
+			 *
+			 * @return name
+			 */
+			virtual const char* get_name() const
```
override

gf712 commented on src/shogun/kernel/MaternKernel.cpp:
```diff
@@ -0,0 +1,85 @@
+
+void CMaternKernel::init()
+{
+	nu = 1.5;
```
move default to declaration

gf712 commented on src/shogun/kernel/MaternKernel.h:
```diff
@@ -0,0 +1,72 @@
+			virtual bool init(CFeatures* l, CFeatures* r);
+
+		protected:
+			float64_t compute(int32_t idx_a, int32_t idx_b);
```
override

gf712 commented on src/shogun/kernel/MaternKernel.h:
```diff
@@ -0,0 +1,72 @@
+			 * @param r features of right-hand side
+			 * @return if initializing was successful
+			 */
+			virtual bool init(CFeatures* l, CFeatures* r);
```
override

gf712 commented on src/shogun/kernel/MaternKernel.cpp:
```diff
@@ -0,0 +1,85 @@
+
+bool CMaternKernel::init(CFeatures* l, CFeatures* r)
+{
+	ASSERT(distance)
```
require?

gf712 commented on src/shogun/kernel/MaternKernel.cpp:
```diff
@@ -0,0 +1,85 @@
+CMaternKernel::~CMaternKernel()
+{
+	cleanup();
+	SG_UNREF(distance);
```
use automatic mem management

gf712 committed:
cleanup + gradient calculation

gf712 commented on src/shogun/kernel/MaternKernel.cpp:
```diff
@@ -0,0 +1,149 @@
+				gradient /= std::pow(width, 2) * std::tgamma(nu);
+
+				return gradient;
+			};
```
@karlnapf as you can see the gradients are pretty complex... I used WolframAlpha to compute them, but would like to test them.. Is there a numeric gradient estimation in shogun I could use?

karlnapf replied:
I vaguely remember that we had something for that somewhere ....

karlnapf replied:
actually, why don't you implement the kernel in jax and generate some simple test cases in there (maybe in a gist that could be put as a comment in the unit test)

gf712 replied:
I can check for matern 1/2, 3/2 and 5/2, but for the arbitrary order there is no way with JAX, as it hasn't implemented bessel of the second kind.. I'll have a look if I can test somewhere else.
Would be good to check the gradients though, like in sklearn unittests for kernels

karlnapf replied:
yeah checking for nontrivial cases already is useful.

karlnapf replied:
I think the finite diff stuff is somewhere in the nn code

gf712 commented on src/shogun/kernel/MaternKernel.h:
```diff
@@ -0,0 +1,78 @@
+	 *		\f$\rho\f$ is the kernel width, and
+	 *		\f$\Gamma\f$ is the gamma function
+	 **/
+	class MaternKernel: public DistanceKernel 
```
@karlnapf do you know from what class I should inherit from?

karlnapf replied:
translation invariant kernel?

gf712 replied:
yea, thought so, will change it

gf712 committed:
gradients and unittests

gf712 commented on tests/unit/kernel/MaternKernel_unittest.cc:
```diff
@@ -0,0 +1,142 @@
+
+TEST(MaternKernelTest_1_2, test_derivative_width)
+{
+	// gradient computed with JAX: https://gist.github.com/gf712/b10ecefd2379e8492c9802ac8cb329c9
```
@karlnapf getting the same results as JAX but different from sklearn..

karlnapf replied:
ha! Is the kernel maybe parametrized differently?

gf712 replied:
I’m not sure.. the kernel matrix is the same, it’s just the derivative that is different

gf712 replied:
Could you actually check the gist please? I think it’s correct.. kernel wrt to width right?

gf712 replied:
It seems like the sklearn gradient is exactly double of what I’m getting..

gf712 replied:
So sklearn overestimates the gradient by a factor of `width`. I tested it numerically, and the JAX/WolframAlpha gradients are correct. I wonder why.. I suspect it is related to this distance calculation
https://github.com/scikit-learn/scikit-learn/blob/95d4f0841d57e8b5f6b2a570312e9d832e69debc/sklearn/gaussian_process/kernels.py#L1437-L1438

karlnapf replied:
maybe it has to do with the fact how they use the gradients downstream

gf712 replied:
maybe, maybe it also doesn't matter too much as long as they are in correct within the same factor.. Anyway, this should be correct

karlnapf commented on src/shogun/kernel/MaternKernel.cpp:
```diff
@@ -0,0 +1,170 @@
+}
+
+MaternKernel::MaternKernel(int32_t size, float64_t w, float64_t order): 
+	ShiftInvariantKernel(), m_nu(order), m_width(w)
```
could you call `MaternKernel` here to avoid having to call register params again?

gf712 replied:
Yup, need to clean this up

karlnapf commented on src/shogun/kernel/MaternKernel.cpp:
```diff
@@ -0,0 +1,170 @@
+
+MaternKernel::MaternKernel(): ShiftInvariantKernel()
+{
+	register_params();
```
I think with initializing parameters in the declaration and calling the base ctor in the other ones, we might not need these methods anymore, could just register directly here?

karlnapf commented on src/shogun/kernel/MaternKernel.cpp:
```diff
@@ -0,0 +1,170 @@
+
+	// first we can check if we should use one of the approximations which are cheaper to calculate
+	// these formulas were taken from Rasmussen's GPs for ML
+	if (Math::fequals(m_nu, 0.5, std::numeric_limits<float64_t>::epsilon()))
```
maybe we should add `allclose` methods to linalg to get rid of this CMath thing?

gf712 replied:
Yes, I can add that in a separate PR and merge that

karlnapf commented on src/shogun/kernel/MaternKernel.cpp:
```diff
@@ -0,0 +1,170 @@
+		result = std::exp(-dist/m_width);
+	else if (Math::fequals(m_nu, 1.5, std::numeric_limits<float64_t>::epsilon()))
+	{
+		float64_t ratio = (std::sqrt(3) * dist) / m_width;
```
is something like sqrt(3) constexpr? :)

gf712 replied:
I think all compilers would precompute this with optimisations on

karlnapf replied:
i see! cool!

gf712 replied:
see line 2 of the assembly in https://godbolt.org/z/xCXF3q. When optimisations are turned off this becomes a runtime call to sqrt

karlnapf commented on tests/unit/kernel/MaternKernel_unittest.cc:
```diff
@@ -0,0 +1,142 @@
+	SGMatrix<float64_t> dMatrix = kernel->get_parameter_gradient(*width);
+
+	for (const auto& [result, reference]: zip_iterator(dMatrix, expected_dmatrix))
+		EXPECT_NEAR(result, reference, 1E-7);		
```
single precision?

gf712 replied:
You mean get epsilon from type traits?

karlnapf replied:
If that is possible I think it should be done or?

gf712 replied:
I am not sure, because the error propagates through each calculation. This is the tolerance value that got everything correct for all the tests without tweaking for each fixture

gf712 commented on src/shogun/kernel/MaternKernel.cpp:
```diff
@@ -0,0 +1,170 @@
+	distance->set_disable_sqrt(false);
+	m_distance=distance;
+	SG_ADD(&m_nu, "nu", "order of the Bessel function of the second kind", ParameterProperties::HYPER)
+	SG_ADD(&m_width, "width", "Kernel scale", ParameterProperties::HYPER)
```
Also GRADIENT

gf712 commented on src/shogun/kernel/MaternKernel.cpp:
```diff
@@ -0,0 +1,197 @@
+					return (std::pow(2, 1 - m_nu) / std::tgamma(m_nu)) *
+					       std::pow(ratio, m_nu) * bessel;
+				};
+				return (k(m_width + epsilon) - k(m_width)) / epsilon;
```
@karlnapf the derivative of the matern kernel is a monster (I also think it is quite unstable outside the widths that are usually used, as I was getting very different result between WolframAlpha gradient and the numerical gradient. So I just put the numerical gradient (like sklearn does)

gf712 replied:
for reference https://www.wolframalpha.com/input/?i=derivative+(2^(1+-+nu)+%2F+Gamma(nu))+*+(sqrt(2+*+nu)+*+d+%2F+w)+^+nu+*+BesselK(nu%2C+sqrt(2+*+nu)+*+d+%2F+w)+wrt+w

karlnapf replied:
ok!

karlnapf replied:
maybe add some comments in the code so that this reasoning is transparent in case someone touches it later

gf712 replied:
sure! I think the main point is that It is more expensive to compute the gradient analytically than numerically. As for the value of the delta in the gradient calculation, it was what gave me the correct values for the width values that we compute analytically (i.e. I computed the 1/2 kernel gradient numerically and this delta gave a good result). I think it might be a bit high though?

gf712 commented on src/shogun/kernel/MaternKernel.cpp:
```diff
@@ -0,0 +1,197 @@
+			    "Shogun compiled on clang does not support Matern kernels with "
+			    "width other than 0.5, 1.5 and 2.5. The current width is {}",
+			    m_width);
+#endif
```
@karlnapf I couldn't find the bessel function I need here (`kv` in scipy) so I think we should just wait until clang implements it. It should also be possible to compile this with clang with libstdc++, rather than libc++

karlnapf replied:
problem will be tests / examples when compiled without clang, but ok

gf712 replied:
I guess, but the option is adding some fortran code from http://netlib.org/amos/

gf712 replied:
I can try and find out how stdlibc++ does it actually

karlnapf replied:
good idea. Not so sure about Fortran code, never did that. Is it easy?

gf712 replied:
I copied the libstdc++ code :D

gf712 commented on src/shogun/kernel/MaternKernel.cpp:
```diff
@@ -0,0 +1,197 @@
+		error(
+		    "Shogun compiled with libc++ does not support Matern kernels with "
+		    "width other than 0.5, 1.5 and 2.5. The current width is {}.",
+		    m_width);
```
it's too complicated to add support for cyl_bessel_k, so will just throw an error when running this part of the code when compiled with libc++. Note that realistically no one would ever use a value other than 0.5, 1.5 and 2.5..

karlnapf commented on src/shogun/mathematics/bessel.cpp:
```diff
@@ -0,0 +1,290 @@
+//
+// This file is part of the GNU ISO C++ Library.  This library is free
+// software; you can redistribute it and/or modify it under the
+// terms of the GNU General Public License as published by the
```
sadly, we cannot do that. It would need to live in shogun-gpl. But let's not add to that. Is there maybe a version in free-domain?

karlnapf replied:
(gpl is viral ... it "infects" our bsd repository with gpl and then our repo also must be gpl)

gf712 replied:
OK, I'll keep looking!

gf712 replied:
is the boost license compatible? I'll try and rewrite their code into something that can work standalone

karlnapf replied:
It is, but then we would need to include the boost license in our license .... gna

karlnapf replied:
> Permission is hereby granted, free of charge, to any person or organization
obtaining a copy of the software and accompanying documentation covered by
this license (the "Software") to use, reproduce, display, distribute,
execute, and transmit the Software, and to prepare derivative works of the
Software, and to permit third-parties to whom the Software is furnished to
do so, all subject to the following:
> The copyright notices in the Software and this entire statement, including
the above license grant, this restriction and the following disclaimer,
must be included in all copies of the Software, in whole or in part, and
all derivative works of the Software, unless such copies or derivative
works are solely in the form of machine-executable object code generated by
a source language processor.

gf712 replied:
I think in that case we should just not have it for libc++. It is quite time consuming to get another implementation, and realistically we will not need it any time soon (this kernel is probably just going to be used for BO)

gf712 replied:
I was looking at how scipy does it, but then we need to add a bunch of fortran code and then translate the scipy C code to something for C++ types. There is a C translation of the main routines but it is but seems very innefficient

karlnapf replied:
yeah ... I had similar issues a while ago when using pdf/cdf functions of the Gamma distribution. I eventually found a public domain version of some old-school code, a packet called cdflib `lib/external/cdflib.cpp`

karlnapf replied:
it is crazy code, full of goto statements ;)

gf712 replied:
this is what I found https://github.com/cran/Bessel/blob/master/src/zbsubs.c. It's like reading Fortran written in C.. It can't be very efficient with so many jumps :p

karlnapf replied:
use it :)

gf712 replied:
OK, I'll have some time during the weekend to clean it up, as this is written to work with Rcpp header files

gf712 commented on src/shogun/kernel/MaternKernel.cpp:
```diff
@@ -0,0 +1,198 @@
+MaternKernel::MaternKernel() : ShiftInvariantKernel()
+{
+	auto distance = std::make_shared<EuclideanDistance>();
+	distance->set_disable_sqrt(false);
```
@karlnapf I turned this off, but I guess it could be true for the sake of efficiency?

karlnapf replied:
what is that for again?

gf712 replied:
It’s to avoid computing the square root of the Euclidean distance

gf712 committed:
added bessel functions

gf712 commented on src/shogun/mathematics/zbsubs.h:
```diff
@@ -0,0 +1,110 @@
+
```
@karlnapf I am not sure what license to put here? This is mostly copied with some slight modifications to get it to compile without any additional libraries

gf712 replied:
there was no license in the repo where I copied this from (probably because it is just C code translated directly from FORTRAN)

karlnapf replied:
It is probably public domain then. I suggest you try a bit to track the license down and if you cannot find anything State the source and that you modified it and say that it’s taken under public domain

gf712 replied:
it seems like they went for GPL... http://r-forge.r-project.org/projects/specfun/
I can just do the f2c myself for the zbsubs routines?

karlnapf replied:
Gpl is non copy pastable
You can do them yourself but not sure how much work that would be?

gf712 replied:
I translated it with f2c, it was some effort, but it should be working fine now

karlnapf replied:
nice one!

gf712 commented on src/shogun/mathematics/bessel.h:
```diff
@@ -0,0 +1,59 @@
+			}
+			else
+			{
+				error("Unexpected error {} returned from zbesk", ierr);
```
These error messages are similar to the R package, but I assume that is not an issue?

karlnapf replied:
I think if there is any chance that a user will face those. They should be improved a tiny bit

gf712 replied:
It is possible, I never had any issues, but I guess if you're using this kernel on any data something can go wrong..

gf712 commented:
@karlnapf I think this is ready to merge now (just waiting for CI)

gf712 commented:
@karlnapf shall I merge this?

karlnapf commented on src/shogun/kernel/MaternKernel.h:
```diff
@@ -0,0 +1,140 @@
+		 * @param nu the order of the bessel function of the second kind
+		 * @param dist distance to be used
+		 */
+		MaternKernel(int32_t size, float64_t width, float64_t nu);
```
I'd suggest to remove this one ... the cache is useless for this kernel, so the size doesnt need to be set

karlnapf commented on src/shogun/kernel/MaternKernel.h:
```diff
@@ -0,0 +1,140 @@
+		 */
+		MaternKernel(
+		    const std::shared_ptr<Features>& l,
+		    const std::shared_ptr<Features>& r, int32_t size, float64_t width,
```
same here, I'd remove the size

karlnapf commented on src/shogun/mathematics/zbsubs.h:
```diff
@@ -0,0 +1,116 @@
+    double* zr, double* zi, double* fnu, int* kode, int* n, double* yr,
+    double* yi, int* nz, int* nlast, double* fnul, double* tol, double* elim,
+    double* alim);
+int zuni2_(
```
tasty :)

karlnapf commented on tests/unit/kernel/MaternKernel_unittest.cc:
```diff
@@ -0,0 +1,190 @@
+	SGMatrix<float64_t> computed_kernel_matrix = kernel->get_kernel_matrix();
+
+	for (const auto & [ result, reference ] :
+	     zip_iterator(computed_kernel_matrix, expected_kernel_matrix))
```
very nice this zip iterator!

gf712 replied:
Indeed! One of these days I need to fix it to work with temporary values, so we can have
```cpp
for (const auto& [X, y]: zip_iterator(DotIterator(feat), label)) {
...
}
```

karlnapf replied:
yes that'd be ace!

karlnapf replied:
maybe make an issue for this? a student might pick it up?

karlnapf commented on tests/unit/kernel/MaternKernel_unittest.cc:
```diff
@@ -0,0 +1,190 @@
+	SGMatrix<float64_t> dMatrix = kernel->get_parameter_gradient(*width_param);
+
+	SGMatrix<float64_t> computed_kernel_matrix = kernel->get_kernel_matrix();
+	kernel->put("width", width + epsilon);
```
class member or c++ setter (as discussed on irc)

gf712 committed:
removed cache size contructor

also added getter/setters

gf712 committed:
fix bug in release mode

gcc 8.4.0 was not happy capturing class members by reference, so copied them into the lambda
